### DISCUSSION TITLE
qpdf 12 planning

### DISCUSSION THREAD

**Discussion by jberkenbilt** [qpdf/qpdf#785] [Ideas]
This discussion is for planning of changes we might want to make for qpdf 12. If a change can be made without breaking binary compatibility (is non-AB-breaking), it doesn't have to wait for qpdf 12, so this is primarily for changes that may break binary compatibility. Any non-backward-compatible API changes necessarily also break ABI, so they can go here too, though we want to avoid API breaking changes whenever possible.

Here is my initial list:
* Consider requiring C++-17 as the minimum version so we can start relying on some C++-17 features. At this time, the most popular C++ compilers default to 17, and the version of g++ in Ubuntu 18.04 supports it, so backporting to LTS versions of most Linux distributions should still work. I think it may be too soon to consider C++-20 given how many people use qpdf in older environments.
* Decide whether private `Members` classes should be `std::unique_ptr` instead of `std::shared_ptr` for classes that can't be copied (like `QPDF`). We would only want to consider this for classes that can't be copied by design, not classes that we may decide sometime to allow copying of.
* @m-holger has ideas for a new alternative to `QPDFObjectHandle` replacement that can interoperate in some way with the existing one but offers a more idiomatic C++ interface that's more consistent with standard library containers and has better type safety

If others have any wishlist items, this would be a good place to track them.

@jbarlow83, @fredemmott -- also calling your attention to this issue

# NOTE

Search for `qpdf 12 TODO` in this discussion for specific things to remember for qpdf 12.

**Comment by jbarlow83** [comment]
C++17 seems like a reasonable change. Mainly it allows std::stringview - there may be further optimization opportunities from the option of using this instead of std::string.

pikepdf currently uses metaclasses to create the illusion of pikepdf.Object being its various actual types. For example pikepdf.Stream is a pikepdf.Object - it just lies about its identity. For example `__repr__` on pikepdf.Object inspects the type code and reports an answer based there on. (pikepdf.Object is QPDFObjectHandle.) The illusion is comprehensive but imperfect and would probably benefit from a more strongly typed C++ class.

**Comment by jberkenbilt** [comment]
Reading over [the Wikipedia article on [C++17](https://en.wikipedia.org/wiki/C%2B%2B17), I see quite a few compelling advantages with `std::stringview` being a big one. Another big one is `std::filesystem`, which would make it possible to implement #553 in a portable way much more easily. I think `std::variant` may have a place too and could reduce use of inheritance in some cases. `__has_include` may make it possible to eliminate some build-time checks made by cmake. Also, I seem to have overlooked the existence of Unicode string literals in C++11, which I could probably have used in some test code.

If switching to C++17 is non ABI-breaking (which I believe will be the case), I may consider requiring it for 11.2 but waiting until 12 before modifying any existing API to use its features.

**Reply by fredemmott** [reply]
> Unicode string literals

These end up being somewhat problematic in practice, along with the rest of the `char8_t` and `u8string` stuff that was added in C++20; the best approaches seems to be to insist that the source code is compiled as UTF8 or use escapes, and use `"normal string literals"`s, `std::string`, and `char`. 

While I agree on C++17 being a better target than C++20 for now, this is still relevant for forward-compatibility, given that `u8"foo"` is a `const char[N]` in C++11-17, but a `const char8_t[N]` in C++20 and above - so using `u8"foo"` now will create issues in the future.

**Comment by m-holger** [comment]
In addition to the compelling advantages, there are lots of little ones that allow for cleaner and more readable code.

**Comment by m-holger** [comment]
It would be useful to have a decision on this sooner rather than later.

**Comment by m-holger** [comment]
> Decide whether private Members classes should be std::unique_ptr instead of std::shared_ptr for classes that can't be copied (like QPDF). We would only want to consider this for classes that can't be copied by design, not classes that we may decide sometime to allow copying of.

IMO, there are two cases to consider: classes that purely provide handles to other objects (like QPDFObjectHandle) and classes that implement objects in their own right (like QPDF).

In the case of handles I do not believe that the handle should have any state other than some kind of pointer to the underlying object, and therefore a private Members class does not make sense to me.

In the case of other objects, I would expect that the result of copying is that we end up with two objects that can be modified independently. This requires that the copy receives a distinct copy of the Members object rather than a pointer to the existing  Members object. Having a unique_ptr therefore makes sense to me.

**Comment by fredemmott** [comment]
>  I think it may be too soon to consider C++-20 given how many people use qpdf in older environments.

I'm a huge fan of C++20, but definitely. It's great when I can say "VS2022 required", but very painful on my cross-platform projects. Even on the latest GCC/clang/apple clang, lots of features are [only partially implemented](https://en.cppreference.com/w/cpp/compiler_support) - it'll be a few years at least until it's broadly available with mature implementations

**Reply by jberkenbilt** [reply]
Indeed -- that's why I went to 11 and 14 even though 17 was there.

**Comment by jberkenbilt** [comment]
Playing around a little, I'm ready to go ahead and switch to require C++17 for 11.2. There are a couple of small fixes (including one I have yet to make) that I'd like to get into 11.1.1 maybe in another week or so, then we can switch it. For linux/gcc, changing from 14 to 17 does cause a few symbols to appear and disappear from the ABI, but they're all things that are not part of the ABI. I'll have to tweak my ABI checking code, and I may want to add a cxx17 test (like the existing cxx11 test) so we get an obvious and quick failure for a feature that's not correctly implemented. I actually found, using that test case, that user-defined string literals with custom matching labels is broken for MSVC 2019.

**Comment by jberkenbilt** [comment]
# qpdf 12 TODO
Remove `#include <qpdf/PointerHolder.hh>` from public header files. They were left there for qpdf 11 so that people who use `PointerHolder` but got its definition implicitly through its indirect inclusion by other headers would see all the warnings and deprecations instead of getting an error that `PointerHolder` wasn't defined. After qpdf 12, we can remove those even if we don't entirely remove `PointerHolder`. We can maybe remove it outright for qpdf 13 or after some adequate amount of calendar time has passed since this release of qpdf 11.

**Reply by jberkenbilt** [reply]
The qpdf 12 release notes should say something like this:

> `<qpdf/PointerHolder.hh>` is no longer included by any public header files. If you are still using `PointerHolder` in your code and have a need to do so for backward compatibility, you must be sure to include the file explicitly where you need to. Note that `PointerHolder` may eventually disappear completely. See ... for information about migrating away from `PointerHolder`

**Reply by jberkenbilt** [reply]
Additionally, with `PointerHolder` no longer included anywhere in the public API, it will no longer be necessary for people who don't include it themselves to define `POINTERHOLDER_TRANSITION`.

**Comment by jberkenbilt** [comment]
# ~~qpdf 12 TODO~~

Consider removing `unsigned char` from the API. While the code has to work whether the default `char` is signed or not, and there may be use cases for `unsigned char` internally, we should only use `char` in the API.

**Reply by jberkenbilt** [reply]
Before deciding whether to do this, we'll have to examine the impact it may have on users.

**Reply by jberkenbilt** [reply]
I think it's not worth removing `unsigned char` from the API. It appears in a lot of places, and changing it to char may introduce bugs. Plus it will be very disruptive for people who have written code that uses the qpdf API and have used unsigned char in their code...it could cause a ripple effect. There are no performance advantages to removing unsigned char. We already have a few places where we have overloaded functions, such as with Pipeline, to make it more convenient.

**Comment by jberkenbilt** [comment]
I have released 11.1.1. I'll give it a week to make sure there are no fires, and then I'll bump the required C++ version up to 17.

**Comment by jbarlow83** [comment]
As an experiment I tried dropping the reference counting hacks from pikepdf with qpdf 11 changes. I expected an explosion of segfaults but everything worked on first attempt, apart from a few test suite failures. These test suite failures were tests that either 1) explicitly tested reference counts to make sure they worked as before, or 2) tests that expected an Object would keep the owner Pdf alive. 

That's a very positive result for the v11 changes - it's easier to bind now and everything has more standard semantics. I was expecting I'd have to add a lot of return value policies to the binding code but none were needed.

I believe this also explains why the move constructor of QPDFObjectHandle didn't work - it violated pikepdf's expectations about reference counts and Objects keeping Pdfs alive. Currently the reference counting hacks prevent a destroyed state QPDFObjectHandle from ever been shown to the user. 

On my end dropping the hacks means Objects no longer keep their parent Pdf alive by holding a reference to it, so when I make this change it will be API breaking. pikepdf users will get exceptions and destroyed Objects if they depend on old behavior. Overall, I think it's better behavior and worth adopting - there's not too much use for keeping an Object around when the Pdf is gone, but I can imagine cases (e.g. extracting an image from a Pdf) where people might have done just that.

When I make the change it should be possible to use QPDFObjectHandle's move constructor.

**Reply by jberkenbilt** [reply]
That's great to hear. Yes, the test failures in pikepdf's test suite with the move constructor were mostly assertion failures around reference counts, but there were also segfaults. I figured out what was going on with the apparent ABI breakage in my testing and fixed my ABI checking script with 11.1.1 (for an unrelated reason), so we don't have to wait for qpdf 12 to put that back in. Do you think we need to coordinate pikepdf 7 with whichever 11.x release adds the QPDFObjectHandle move constructor, or will it work for you to make your changes and release a new version and then the qpdf change to follow? Since the move constructor was just a couple of lines in the header, it would be easy for you to test with and without the existence of that constructor.

I guess I will need to call attention to the addition of the move constructor in the header since it could break other people's assumptions about reference counts as well, if anyone else cares. That might be a reason to wait for qpdf 12.

**Reply by jbarlow83** [reply]
I confirmed that my new branch works fine with move constructors put back into QPDF.

It's probably better to wait for qpdf 12 to introduce it, and I'll try to slip in pikepdf 7 before then.

**Reply by jberkenbilt** [reply]
Sounds good.

# qpdf 12 TODO

Reintroduce move constructors for `QPDFObjectHandle`

**Reply by m-holger** [reply]
@jbarlow83 Do you have any time frame for pikepdf 7?

@jberkenbilt  Should the upcoming change be mentioned in the release notes for 11.3?

**Reply by jberkenbilt** [reply]
We don't need to mention anything about qpdf 12 in the release notes for 11.3 unless there is some action people can/should take to prepare, and for this list, I don't think there is anything.

**Reply by jbarlow83** [reply]
pikepdf 7 can probably be released in a few days.

**Reply by jbarlow83** [reply]
> @jbarlow83 Do you have any time frame for pikepdf 7?
> 


I released pikepdf 7 today. I was able to drop almost all of my shared_ptrs and replace them with unique_ptr - which is much cleaner when Python is reference counting an object known to both.

Some quirky situations like a Stream keeping a stale QPDF alive and returning empty if the QPDF is closed now raise an exception. That's also better behavior. 

I don't think it's worth proceeding with my earlier suggestion of having a fixed smart pointer container QPDF. There are many benefits to being able to control how an object is allocated.

**Comment by jberkenbilt** [comment]
# qpdf 12 TODO

Clarify in the documentation/release notes that, for `--pages`, the default behavior is to preserve any supported document-level features. Prior to the pages epic work, this is only page labels. In the future, it will be possible to opt in or out of preservation of document-level structures, but the default is that, as new ones are added to qpdf, they will be considered by default, even if that slows down page operations.

**Comment by m-holger** [comment]
# qpdf 12 TODO
Pass QPDFObjGen s by value instead of const&.

**Reply by jberkenbilt** [reply]
Agreed.
```
    std::cout << sizeof(QPDFObjGen) << " " << sizeof(QPDFObjGen const&) << std::endl;
```
shows both have a size of 8 on my platform. What would be the advantage of passing by value instead of `const&`? I guess there is a non-zero cost to taking the address of something and dereferencing it. In that case, we should probably always pass small things by value. I notice that `std::string_view` is 16 bytes where `std::string const&` is 32. This must be because of inheritance, so I guess std::string_view is definitely cheaper to pass around than `std:string const&`. I guess I have always treated passing a reference as cheaper than passing a value as long as the value at least the size of a pointer, but I guess that's not quite accurate.

**Reply by m-holger** [reply]
Without claiming any expertise, my understanding is that it is idiomatic to pass small (say <= 2 words), simple (eg no virtual destructors) by value. This includes string_views. Reasons include that it may be more efficient because it is more likely that the parameters can be passed in registers rather than on the stack, avoids unnecessary indirection and is easier for the compiler to optimize as a const& may be aliased by another variable, while a copy can't.

In the case of QPDFObjGens, passing by const& has caused problems in the past where the referenced variable changed within the body of the method in ways that were difficult to trace.

**Reply by jberkenbilt** [reply]
Right, that makes sense. I am in full agreement with the proposal.

**Reply by m-holger** [reply]
There is the question whether the change should exclude virtual methods that users are expected to override, such as StreamDataProvider::provideStreamData.

My view would be that the benefit does not justify the disruption to users.

**Reply by jberkenbilt** [reply]
Yeah, it would be great if we can do this in a way that doesn't break API (only ABI) so that no code changes are required, especially since qpdf 11 made people make code changes for PointerHolder.

**Comment by jberkenbilt** [comment]
Note about `std::string_view`:

I'm thinking we should probably never return `std::string_view`, and we should definitely never replace a return type of `std::string const&` with a return type of `std::string_view`. The reason is that, if `f()` returns `std::string const&`, then `auto x = f()` results in `x` being a `std::string`, but if `f()` returns `std::string_view`, then `x` is a `std::string_view`. With no compilation warning, code goes from safe to unsafe. We should never do that. For example, when I run this (piped to cat -v):

```c++
#include <iostream>
#include <string>
#include <string_view>

class A
{
  public:
    std::string const& f() {
        return s;
    }
    std::string_view g() {
        return s;
    }
    void set(std::string_view v) {
        s = v;
    }

  private:
    std::string s;
};

int main()
{
    A a;
    a.set("three");
    auto x1 = a.f();
    auto y1 = a.g();
    std::cout << x1 << std::endl;
    std::cout << y1 << std::endl;
    a.set("two");
    auto x2 = a.f();
    auto y2 = a.g();
    std::cout << x1 << " " << x2 << std::endl;
    std::cout << y1 << " " << y2 << std::endl;
    return 0;
}
```
I get the output:
```
three
three
three two
two^@e two
```

And this:

```c++
#include <iostream>
#include <string>
#include <string_view>
#include <memory>

class A
{
  public:
    std::string const& f() {
        return *s;
    }
    std::string_view g() {
        return *s;
    }
    void set(std::string_view v) {
        s = std::make_shared<std::string>(v);
    }

  private:
    std::shared_ptr<std::string> s;
};

int main()
{
    A a;
    a.set("three");
    auto x1 = a.f();
    auto y1 = a.g();
    std::cout << x1 << std::endl;
    std::cout << y1 << std::endl;
    a.set("two");
    auto x2 = a.f();
    auto y2 = a.g();
    std::cout << x1 << " " << x2 << std::endl;
    std::cout << y1 << " " << y2 << std::endl;
    return 0;
}
```
when compiled with `clang++ -fsanitize=address` generates a memory error.

While it would be okay to return a `std::string_view` to a static, it creates a risk of introducing future errors. While returning a std::string_view that might go bad with appropriate disclaimer may be acceptable in some performance-critical places, it would not be okay if the function previously returned a `std::string const&` because of the behavior with `auto`.

**Reply by jbarlow83** [reply]
It's fine to return `constexpr std::string_view` and that is probably what is wanted for a static. Then `auto` picks up the `constexpr` and keeps you out of trouble.

**Reply by jberkenbilt** [reply]
Ah, excellent. I did not know about that. Of course, it doesn't help with an example like the above, but it would make it safe in cases where constexpr works. Thanks!

**Comment by m-holger** [comment]
Consider guaranteeing that no method in the public API will return an unresolved object handle.

Make methods such QPDFObjectHandle:: isArray etc affected by such guarantee const / const noexcept

**Comment by jberkenbilt** [comment]
I'm okay with setting the minimum version of C++ for qpdf 12 to C++20. My rule of thumb is that I use whatever C++ version is supported in the oldest supported Ubuntu LTS release. Ubuntu 20.04 drops out of security support in April 2025, but because of the 64-bit time_t issue and dependency on particular debian infrastructure packages, backporting any qpdf > 11.9.1 to Ubuntu 20.04 is a lot of extra work. As such, I'm not backporting 12 to my PPA for 20.04 regardless of whether it's out before 20.04 drops from support. For that reason, I'm fine with C++ 20. That enables us to use std::filesystem, which means we can use an attachment's modification time if we want when adding it. There may be a few other C++ 20 comments in the code. (There's a reference to using starts_with in QUtil.cc.)

**Reply by jbarlow83** [reply]
I agree with this change and see no issue for pikepdf. It means I will do a major version bump and change to manylinux 2.28. 

Whenever people have asked me to renew support for older versions after dropping them, I say I'd be happy to accept PRs if they're willing to take on the burden of maintaining support for old versions, or if they're willing to pay me to do it. So far, I've found no one who wants legacy support enough to put their money where their mouth is. As open source developers our time is scarce and the value of our contributions often underappreciated. Cutting old versions lets us take advantage of new features in standard libraries, saving precious time. Anyway, I'll get off the soapbox. This is a good move for qpdf.

**Comment by jberkenbilt** [comment]
We're going to give qpdf 11.10 a week or so to see if there are any fires and then open an ABI window for qpdf 12. I'm hoping to get qpdf 12 out by March, but if I can't, I won't. That means we will focus our efforts. Some of the items above have already been done in the 11 series. I am definitely going to do the PointerHolder change (making the default be that PointerHolder is not available; you can still get it by defining POINTERHOLDER_TRANSITION).

The decision about when to move to C++-20 doesn't have to be made in tandem with going to qpdf 12. We actually bumped to C++-17 with version 11.2.0. I'm strongly inclined to go to C++-20 slightly after the release of qpdf 12, maybe in a 12.1 release. That way, qpdf 12 will be able to be backported to more systems. At this time, C++-20 support is virtually complete in gcc 11, which is the version in Ubuntu 22.04. That will be the oldest supported Ubuntu LTS by the end of April. I already don't plan on backporting qpdf 12 to focal (Ubuntu 20.04) in my PPA because I can't build the doc packages there, and I skipped 11.10 for Jammy (22.04) because of the 64-bit time_t issue in debian-derived distributions. With qpdf 12, I'll be bumping the library soname and can break from the t64 stuff as a special case, at which point I'll be able to backport to Jammy again without jumping through hoops.

Once we're comfortable with 11.10, it is my intention to merge the PointerHolder changes (not technically ABI breaking, but a source-compatibility change that could break some code, so worthy of a major release) and to go through the ABI comments in the code making changes that we had to delay until an ABI window. There are a lot of potential ABI changes that are not quite ready, so maybe 12 will be a relatively short release and we'll do a 13 before the end of the year...hard to say. I don't like doing major release too often because it's disruptive for distributions and packagers, but the time between 11 and 12 is on the long end of the spectrum.

Anyone following should feel free to advocate for ABI-breaking changes that are ready to go in or to raise concerns about going to C++-20. @jbarlow83 has already weighed in that C++-20 would be welcome. There are a few features that are blocked by it, such as reading file system times in a portable way....not that I couldn't have implemented it in some other way of course.

My brain is still more or less on C++-14. I plan to make an effort to bring myself up-to-date so I can make better use of newer language features, though I'm not the main person making code changes right now.

**Comment by m-holger** [comment]
The release of qpdf 12.0 is currently planned for the weekend of 9 March 2025.

A summary of the changes included can be found in the [release notes](https://qpdf.readthedocs.io/en/latest/release-notes.html).

**Reply by jberkenbilt** [reply]
Note that qpdf 12.0 will still require C++-17. This is a tactical move so that people who are stuck on 17 can benefit from many of the API improvements for 12.0. We will bump the required C++ version to 20 in 12.1 or later, but fairly soon after 12.0.

**Comment by rubyFeedback** [comment]
We are now at 12.x, soon 12.1.1. :)

https://github.com/qpdf/qpdf/commit/cc933c35f6f2164c306231778b8598c6a7eccb4a

The discussion here started in late 2022 for qpdf 12. Should we have a new thread for
qpdf 13? I don't have any "pressing needs", mind you. Probably the only thing I would
suggest is the documentation + commandline examples, so mostly documentation
related; but I feel that this is also far from very high priority, just something that may
be nice to have, e. g. "do everything related to .pdf files". And perhaps plug-ins like
converting .epub to .pdf (I still prefer .pdf even if they have a reduced functionality
compared to .epub files; calibre is a bit unwieldy and clunky in my opinion, but I
understand that qpdf's primary use case is for .pdf files).
